sctp/src/association/association_internal.rs

```diff
@@ -1894,70 +1894,68 @@ impl AssociationInternal {
-                    sis_to_reset.push(stream_identifier);
-                    if self
-                        .pending_queue
-                        .pop(beginning_fragment, unordered)
-                        .await
-                        .is_none()
-                    {
-                        log::error!("failed to pop from pending queue");
-                    }
-                    continue;
-                }
+        if self.pending_queue.len() == 0 {
+            return (chunks, sis_to_reset);
+        }
 
-                if self.inflight_queue.get_num_bytes() + data_len > self.cwnd as usize {
-                    break; // would exceeds cwnd
-                }
+        // RFC 4960 sec 6.1.  Transmission of DATA Chunks
+        //   A) At any given time, the data sender MUST NOT transmit new data to
+        //      any destination transport address if its peer's rwnd indicates
+        //      that the peer has no buffer space (i.e., rwnd is 0; see Section
+        //      6.2.1).  However, regardless of the value of rwnd (including if it
+        //      is 0), the data sender can always have one DATA chunk in flight to
+        //      the receiver if allowed by cwnd (see rule B, below).
+        while let Some(c) = self.pending_queue.peek() {
+            let (beginning_fragment, unordered, data_len, stream_identifier) = (
+                c.beginning_fragment,
+                c.unordered,
+                c.user_data.len(),
+                c.stream_identifier,
+            );
 
-                if data_len > self.rwnd as usize {
-                    break; // no more rwnd
+            if data_len == 0 {
+                sis_to_reset.push(stream_identifier);
+                if self
+                    .pending_queue
+                    .pop(beginning_fragment, unordered)
+                    .is_none()
+                {
+                    log::error!("failed to pop from pending queue");
                 }
+                continue;
+            }
 
-                self.rwnd -= data_len as u32;
+            if self.inflight_queue.get_num_bytes() + data_len > self.cwnd as usize {
+                break; // would exceed cwnd
+            }
 
-                if let Some(chunk) = self
-                    .move_pending_data_chunk_to_inflight_queue(beginning_fragment, unordered)
-                    .await
-                {
-                    chunks.push(chunk);
-                }
+            if data_len > self.rwnd as usize {
+                break; // no more rwnd
             }
 
-            // the data sender can always have one DATA chunk in flight to the receiver
-            if chunks.is_empty() && self.inflight_queue.is_empty() {
-                // Send zero window probe
-                if let Some(c) = self.pending_queue.peek().await {
-                    let (beginning_fragment, unordered) = (c.beginning_fragment, c.unordered);
+            self.rwnd -= data_len as u32;
```

Comment:
Just a thought that struck me, can this underflow? Out of an abundance of caution how would we handle this possibility pre-emptively?

Reply:
No, we cannot because we check that `self.rwnd` is big enough

```rust
if data_len > self.rwnd as usize {
   break; // no more rwnd
}
```

https://github.com/webrtc-rs/webrtc/blob/39e8427c5917985bcbaaae907d796b66263a0618/sctp/src/association/association_internal.rs#L1936-L1938

Reply:
Cool 👍🏼